cpp/src/arrow/compute/api_aggregate.h

```diff
@@ -142,6 +142,20 @@ Result<Datum> MinMax(const Datum& value,
                      const MinMaxOptions& options = MinMaxOptions::Defaults(),
                      ExecContext* ctx = NULLPTR);
 
+/// \brief Test whether any element in a boolean array evaluates to true.
+///
+/// This function returns true if any of the elements in the array evaluates
+/// to true and false otherwise. Null values are taken to evaluate to false.
```

Comment:
Is this the behaviour we want? (the *"null values are taken to evaluate to false."*

Any/all are of course not the most typical reductions (so I am also not fully sure about the desired behaviour), but, for other reductions we actually skip nulls. And skipping nulls is not the same as evaluating them to False

(somewhat related to my comment at https://github.com/apache/arrow/pull/8294#issuecomment-701357437)

Reply:
Ah, yes, you're correct that for this PR there is not yet a difference: you are only dealing with `any`, and it's only for `all` that there is a difference between "skipping nulls" and "treating nulls as False"

So I would indeed update the docstring to simply state that nulls are skipped.

For your example about `all` (I assume the code was meant to use `all` and not `any` ?), for the last line about Kleene logic I expect a return value of `null` (as the `ǹull` in the values can be both True or False, meaning that the result can be True of False, meaning it is unknown)

Reply:
Docstring updated.

Sorry yes. The examples were for `all`. Just to make sure I understand, with kleene logic, we emit null if there's a null anywhere in the input
```
any_kleene([true, null]) = null
any_kleene([false, null]) = null
all_kleene([true, null]) = null
all_kleene([false, null]) = null
```
I opened ARROW-10301 re: `all` kernel

Reply:
> Sorry yes. The examples were for all. Just to make sure I understand, with kleene logic, we emit null if there's a null anywhere in the input

No, we only emit null if the presence of the null (the fact that it could be either True or False) would influence the result:

```
any_kleene([true, null]) = true
any_kleene([false, null]) = null
all_kleene([true, null]) = null
all_kleene([false, null]) = false
```

But, the above is only when *not* skipping nulls (because with the default of skipping, there is no difference with non-kleene logic)

See the links I mentioned in https://github.com/apache/arrow/pull/8294#issuecomment-701357437, and also the Julia docs have a good explanation of three-valued (Kleene) logic: https://docs.julialang.org/en/v1/manual/missing/index.html#Logical-operators-1

Reply:
BTW, the result you show (a null as result in all cases) is what I expect for the *non*-kleene version when not skipping nulls (I would expect that nulls propagate in that case, and not necessarily be interpreted as false)